File: frontend/src/engine/capabilities/topN.ts
/**
 * topN – Extract the top N or bottom N rows from a range sorted by a value column.
 *
 * Reads the full data range, sorts by the specified value column, takes the
 * first N rows, and writes them to the output range.
 */

import { CapabilityMeta, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "topN",
  description: "Extract the top N or bottom N rows from a range sorted by a value column",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: any,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { dataRange, valueColumn, n, position = "top", outputRange, hasHeaders = true } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would extract ${position} ${n} rows from ${dataRange} by column ${valueColumn} to ${outputRange}` };
  }

  options.onProgress?.("Reading data...");

  const rng = resolveRange(context, dataRange);
  rng.load("values");
  await context.sync();

  const vals = (rng.values ?? []) as (string | number | boolean | null)[][];
  if (vals.length < 2) {
    return { stepId: "", status: "success", message: "Not enough rows." };
  }

  // Separate header and data rows
  const headerRow = hasHeaders ? vals[0] : null;
  const dataRows = hasHeaders ? vals.slice(1) : vals.slice();
  const colIdx = valueColumn - 1;

  options.onProgress?.(`Sorting by column ${valueColumn} (${position})...`);

  // Sort data rows
  dataRows.sort((a, b) => {
    const aVal = Number(a[colIdx]) || 0;
    const bVal = Number(b[colIdx]) || 0;
    return position === "top" ? bVal - aVal : aVal - bVal;
  });

  // Take first N rows
  const taken = dataRows.slice(0, n);

  // Build output
  const output: (string | number | boolean | null)[][] = [];
  if (headerRow) {
    output.push(headerRow);
  }
  output.push(...taken);

  options.onProgress?.(`Writing ${taken.length} rows to ${outputRange}...`);

  const outRng = resolveRange(context, outputRange);
  outRng.getResizedRange(output.length - 1, (output[0]?.length ?? 1) - 1).values = output as any;
  await context.sync();

  const label = position === "top" ? "Top" : "Bottom";
  return {
    stepId: "",
    status: "success",
    message: `${label} ${taken.length} rows by column ${valueColumn} written to ${outputRange}`,
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/top_n.py
"""topN — extract the top N (or bottom N) rows sorted by a value column.

Reads the full data range, sorts rows by the specified 1-based value column
descending (for 'top') or ascending (for 'bottom'), takes the first N rows,
and writes them — with optional header row preserved — to the output range.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def _to_float(x: Any) -> float:
    try:
        if x is None or x == "":
            return 0.0
        return float(x)
    except (TypeError, ValueError):
        return 0.0


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    data_range = params.get("dataRange")
    value_column = params.get("valueColumn")
    n = params.get("n")
    position = (params.get("position") or "top").lower()
    output_range = params.get("outputRange")
    has_headers = bool(params.get("hasHeaders", True))

    if not data_range or not output_range:
        return {
            "status": "error",
            "message": "topN requires 'dataRange' and 'outputRange'.",
        }
    if value_column is None or n is None:
        return {
            "status": "error",
            "message": "topN requires 'valueColumn' and 'n'.",
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": (
                f"Would extract {position} {n} rows from {data_range} "
                f"by column {value_column} to {output_range}."
            ),
        }

    try:
        rng = resolve_range(ctx.workbook_handle, data_range)
        raw = rng.value
        if raw is None:
            return {"status": "success", "message": "Not enough rows.", "outputs": {}}
        if not isinstance(raw, list):
            vals: list[list[Any]] = [[raw]]
        elif raw and not isinstance(raw[0], list):
            rows_, cols_ = rng.shape
            vals = [list(raw)] if rows_ == 1 else [[v] for v in raw]
        else:
            vals = [list(r) for r in raw]

        if len(vals) < 2:
            return {"status": "success", "message": "Not enough rows.", "outputs": {}}

        header_row = vals[0] if has_headers else None
        data_rows = list(vals[1:]) if has_headers else list(vals)
        col_idx = int(value_column) - 1

        # Sort data rows — 'top' = descending, 'bottom' = ascending.
        data_rows.sort(
            key=lambda row: _to_float(row[col_idx] if col_idx < len(row) else 0),
            reverse=(position == "top"),
        )

        taken = data_rows[: int(n)]

        output: list[list[Any]] = []
        if header_row is not None:
            output.append(header_row)
        output.extend(taken)

        if not output:
            return {"status": "success", "message": "No rows to write.", "outputs": {}}

        rows = len(output)
        cols = max((len(r) for r in output), default=0)
        out_rng = resolve_range(ctx.workbook_handle, output_range)
        out_rng.resize(rows, cols).value = output
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"topN failed: {exc}", "error": str(exc)}

    label = "Top" if position == "top" else "Bottom"
    return {
        "status": "success",
        "message": f"{label} {len(taken)} rows by column {value_column} written to {output_range}.",
        "outputs": {"outputRange": output_range},
    }


registry.register("topN", handler, mutates=True)
